### PR TITLE
feat(react-tree): creates SubtreeContext to avoid spreading root context

### PR DESCRIPTION
**Pull request by bsunderhus**
## Previous Behavior

Every single subtree level of a tree would re-declare the whole tree context, causing unnecessary re-renders on the subtrees whenever an item was opened or checked

## New Behavior

Splits root context and subtree context, now if a tree item wants to know if it's opened or checked, it'll get that information directly from the root context, without causing any re-renders on the subtree between them.

This will reduce the amount of re-renders in the subtrees improving performance on nested scenarios.

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">74</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount_base.html">14</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount.html">12</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Avatar-mount_base.html">618</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Avatar-mount.html">591</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Button-mount_base.html">324</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Button-mount.html">323</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Field-mount_base.html">1112</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Field-mount.html">1138</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProvider-mount_base.html">704</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProvider-mount.html">707</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-mount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-mount.html">90</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">72</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">68</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">74</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount_base.html">14</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount.html">12</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/MakeStyles-mount_base.html">863</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/MakeStyles-mount.html">852</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Persona-mount_base.html">1780</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/Persona-mount.html">1683</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/SpinButton-mount_base.html">1378</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29194/perf-test-react-components/SpinButton-mount.html">1356</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                                                                             |      Baseline (minified/GZIP) |                            PR |                                                                                                                                                                                                  Change |
| :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: | ----------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `208.393 kB`<br />`59.395 kB` | `208.392 kB`<br />`59.385 kB` | `-1 B` <br />`-10 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                |         Size (minified/GZIP) |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.572 kB`<br />`19.658 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>               | `40.966 kB`<br />`13.569 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                            |   `6.541 kB`<br />`2.227 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/48153fef6aa6ba63abfdfad25fb76dea13395b94'>48153fef6aa6ba63abfdfad25fb76dea13395b94</a></sub>
<div id="bundleSizeComment9423"/>

File: change/@fluentui-react-components-79050853-fe98-49ee-994b-b9dca0dd661a.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-components-79050853-fe98-49ee-994b-b9dca0dd661a.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=29194&commitId=2880047a562d2ee595753d562550b5789222ab17&env=prod&clientType=FLUENTUI)


  <details><summary><b>Tree</b> 8 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Tree.default open tree.chromium.png|871| Changed|
|Tree.default open tree - RTL.chromium.png|876| Changed|
|Tree.default open tree - Dark Mode.chromium.png|791| Changed|
|Tree.default open tree - High Contrast.chromium.png|1174| Changed|
|Tree.persona - Dark Mode.chromium.png|492| Changed|
|Tree.persona - High Contrast.chromium.png|901| Changed|
|Tree.persona.chromium.png|653| Changed|
|Tree.persona - RTL.chromium.png|653| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-tree/src/contexts/subtreeContext.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  contextType: 'subtree',
+};
+
+/**
+ * @internal
+ */
+export const defaultRootTreeContextValue: SubtreeContextValue = {
+  level: 1,
+  contextType: 'subtree',
+};
```
Comment (layershifter) [inline packages/react-components/react-tree/src/contexts/subtreeContext.ts:22]: There should a single default value for the context, please remove it.

File: packages/react-components/react-tree/src/components/TreeProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
+import {
+  SubtreeContext,
+  SubtreeContextValue,
+  TreeContextValue,
+  TreeContext,
+  defaultRootTreeContextValue,
+} from '../contexts';
+
+export const TreeProvider = React.memo((props: React.ProviderProps<TreeContextValue | SubtreeContextValue>) => {
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/TreeProvider.tsx:10]: `.memo()` there is useless as you have `children` and they will never have a stable reference

File: packages/react-components/react-tree/src/components/TreeProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  if (props.value.contextType === 'subtree') {
+    return <SubtreeContext.Provider value={props.value}>{props.children}</SubtreeContext.Provider>;
+  }
+  return (
+    <TreeContext.Provider value={props.value}>
+      <SubtreeContext.Provider value={defaultRootTreeContextValue}>{props.children}</SubtreeContext.Provider>
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/TreeProvider.tsx:16]: You can have this as a constant in a file, but please don't call it `defaultRootTreeContextValue` 🐱

File: packages/react-components/react-tree/src/contexts/treeContext.ts [hunk 5]
```diff
@@ -25,12 +26,19 @@ export type TreeItemRequest = { itemType: TreeItemType } & (
...
+ * should work exactly the same as Omit
+ */
 type OmitWithoutExpanding<P, K extends string | number | symbol> = P extends unknown ? Omit<P, K> : P;
 
-const defaultContextValue: TreeContextValue = {
-  treeType: 'nested',
+/**
+ * @internal
+ */
+export const defaultTreeContextValue: TreeContextValue = {
```
Comment (layershifter) [inline packages/react-components/react-tree/src/contexts/treeContext.ts:38]: Is export is needed there?

File: packages/react-components/react-tree/src/contexts/subtreeContext.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+export type SubtreeContextValue = {
+  contextType: 'subtree';
+  level: number;
+};
+
+/**
+ * @internal
+ */
+export const defaultSubTreeContextValue: SubtreeContextValue = {
```
Comment (layershifter) [inline packages/react-components/react-tree/src/contexts/subtreeContext.ts:11]: This probably should not be exported

File: packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts [hunk 7]
```diff
@@ -1,22 +1,41 @@
...
+      contextType: type,
+      level,
+      requestTreeResponse,
+    };
 
-  return { tree };
+    return { tree };
+  }
+  return {
+    // level is statically determined by the context
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts:37]: the comment is not useful there as the condition is based on `state.contextType`

File: packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts [hunk 8]
```diff
@@ -1,22 +1,41 @@
...
-    appearance,
-    checkedItems,
-    selectionMode,
-    requestTreeResponse,
-  };
+  if (state.contextType === 'root') {
+    const {
+      openItems,
+      level,
+      contextType: type,
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts:10]:
```suggestion
      contextType,
```

File: packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts [hunk 9]
```diff
@@ -1,22 +1,41 @@
...
+     * there is no sense to memoize it
+     */
+    const tree: TreeContextValue = {
+      treeType,
+      size,
+      openItems,
+      appearance,
+      checkedItems,
+      selectionMode,
+      contextType: type,
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/Tree/useTreeContextValues.ts:29]:
```suggestion
      contextType,
```

File: packages/react-components/react-tree/src/components/FlatTree/useFlatTreeContextValues.ts [hunk 10]
```diff
@@ -1,5 +1,33 @@
...
+   * there is no sense to memoize it
+   */
+  const tree: TreeContextValue = {
+    treeType,
+    size,
+    openItems,
+    appearance,
+    checkedItems,
+    selectionMode,
+    contextType: type,
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/FlatTree/useFlatTreeContextValues.ts:27]:
```suggestion
    contextType,
```

File: packages/react-components/react-tree/src/components/FlatTree/useFlatTreeContextValues.ts [hunk 11]
```diff
@@ -1,5 +1,33 @@
...
+import type { TreeContextValue } from '../../contexts';
+import type { FlatTreeContextValues, FlatTreeState } from './FlatTree.types';
 
-export const useFlatTreeContextValues_unstable: (state: FlatTreeState) => TreeContextValues =
-  useTreeContextValues_unstable;
+export const useFlatTreeContextValues_unstable = (state: FlatTreeState): FlatTreeContextValues => {
+  const {
+    openItems,
+    level,
+    contextType: type,
```
Comment (layershifter) [inline packages/react-components/react-tree/src/components/FlatTree/useFlatTreeContextValues.ts:8]:
```suggestion
    contextType,
```

File: packages/react-components/react-tree/src/contexts/treeContext.ts [hunk 12]
```diff
@@ -6,8 +6,9 @@ import { ImmutableMap } from '../utils/ImmutableMap';
 import { TreeCheckedChangeData, TreeNavigationData_unstable, TreeOpenChangeData } from '../Tree';
 
 export type TreeContextValue = {
-  treeType: 'nested' | 'flat';
+  contextType: 'root';
```
[the remark this answers is not in the document]
Comment (bsunderhus) [reply packages/react-components/react-tree/src/contexts/treeContext.ts:9]: 🤡

File: packages/react-components/react-tree/src/contexts/treeContext.ts (hunk 5 above)
Comment (bsunderhus) [reply packages/react-components/react-tree/src/contexts/treeContext.ts:38]: This is being spread in the `useSubFlatTree` state return, so yes, this has to be exported.

**Review by layershifter — APPROVED** [review]

**Pull request merged**